src/hooks/useFetchExperience.ts

```diff
@@ -0,0 +1,173 @@
+import type { ContentfulClientApi, Entry } from 'contentful'
+import { useCallback, useState } from 'react'
+import { fetchExperienceEntities, fetchExperienceEntry } from '../core/fetchers'
+import { Experience, ExternalSDKMode } from '../types'
+import { createExperience } from '../core'
+
+const errorMessagesWhileFetching = {
+  experience: 'Failed to fetch experience',
+  experienceReferences: 'Failed to fetch entities, referenced in experience',
+}
+
+const handleError = (generalMessage: string, error: unknown) => {
+  const message = error instanceof Error ? error.message : `Unknown error: ${error}`
+  console.error(`${generalMessage} with error: ${message}`)
+}
+
+type useClientsideExperienceFetchersProps = {
+  mode: ExternalSDKMode
+  client: ContentfulClientApi<undefined>
+}
+
+export const useFetchExperience = ({ mode, client }: useClientsideExperienceFetchersProps) => {
+  const [experience, setExperience] = useState<Experience | undefined>(undefined)
+  const [isFetching, setIsFetching] = useState(false)
+
+  /**
+   * Fetch experience entry using slug as the identifier
+   * @param {string} experienceTypeId - id of the content type associated with the experience
+   * @param {string} slug - slug of the experience (defined in entry settings)
+   * @param {string} localeCode - locale code to fetch the experience. Falls back to the currently active locale in the state
+   */
+  const fetchBySlug = useCallback(
+    async ({
+      experienceTypeId,
+      slug,
+      localeCode,
+    }: {
+      experienceTypeId: string
+      slug: string
+      localeCode: string
+    }): Promise<Experience | undefined> => {
+      setIsFetching(true)
+
+      let experienceEntry: Entry | undefined = undefined
+
+      try {
+        experienceEntry = await fetchExperienceEntry({
+          client,
+          experienceTypeId,
+          locale: localeCode,
+          identifier: {
+            slug,
+          },
+        })
+      } catch (e) {
+        handleError(errorMessagesWhileFetching.experience, e)
+        setIsFetching(false)
+        throw e
+      }
+
+      if (!experienceEntry) {
+        const error = new Error(`No experience entry with slug: ${slug} exists`)
+        handleError(errorMessagesWhileFetching.experience, error)
+        setIsFetching(false)
+        throw error
+      }
+
+      try {
+        const { entries, assets } = await fetchExperienceEntities({
+          client,
+          experienceEntry,
+          locale: localeCode,
+        })
+
+        const experience = createExperience({
+          experienceEntry,
+          referencedAssets: assets,
+          referencedEntries: entries,
+          locale: localeCode,
+          mode,
+        })
+
+        setExperience(experience)
+        setIsFetching(false)
+
+        return experience
+      } catch (e) {
+        handleError(errorMessagesWhileFetching.experienceReferences, e)
+        setIsFetching(false)
+        throw e
+      }
```

Comment:
You can basically move all into one single try/catch/finally block like this. Personally, I think it's now easier to follow the things being executed in the function as the critical logic is all happening inside one `try` block with no differing indentations.

```suggestion
      setIsFetching(true)
      try {
        const experienceEntry = await fetchExperienceEntry({
          client,
          experienceTypeId,
          locale: localeCode,
          identifier: {
            slug,
          },
        })
         if (!experienceEntry) {
           throw  new Error(`No experience entry with slug: ${slug} exists`)
         }
         const { entries, assets } = await fetchExperienceEntities({
          client,
          experienceEntry,
          locale: localeCode,
        })
        const experience = createExperience({
          experienceEntry,
          referencedAssets: assets,
          referencedEntries: entries,
          locale: localeCode,
          mode,
        })
        setExperience(experience)
        return experience
      } catch (error) {
        handleError(errorMessagesWhileFetching.experience, error)
        throw error
      } finally {
        setIsFetching(false)
      }
```

Reply:
But then there is a question - if line 53 works, will `isFetching` be `true` or `false`

Reply:
It will be `false`. `Finally` is always called and that's why it's perfect for this kind of "clean-ups" / "resets"

Reply:
I noticed that my proposal didn't cover the different error messages. I just updated it with these two lines: 
```
const errorMessageKey = experienceEntry ? 'experienceReferences' : 'experience'
handleError(errorMessagesWhileFetching[errorMessageKey], error)
```
So now it does exactly the same as above with differing error messages :)

Reply:
Validated. It indeed still triggers `finally`. Well good, then this simplifies things. At first I actually had it just like you described, but then I started doubting whether finally would get triggered even after the error is thrown within `catch`

<img width="926" alt="Screenshot 2023-10-24 at 11 54 55" src="https://github.com/contentful/experience-builder/assets/4171202/5328a56f-0e17-4c1c-91bd-fa6854f21d52">